mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala

```diff
@@ -0,0 +1,238 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.regression
+
+import java.io.Serializable
+import java.util.Arrays.binarySearch
+
+import org.apache.spark.api.java.{JavaDoubleRDD, JavaRDD}
+import org.apache.spark.rdd.RDD
+
+/**
+ * Regression model for Isotonic regression
+ *
+ * @param features Array of features.
+ * @param labels Array of labels associated to the features at the same index.
+ */
+class IsotonicRegressionModel (
+    features: Array[Double],
+    val labels: Array[Double])
+  extends Serializable {
+
+  /**
+   * Predict labels for provided features
+   * Using a piecewise constant function
+   *
+   * @param testData features to be labeled
+   * @return predicted labels
+   */
+  def predict(testData: RDD[Double]): RDD[Double] =
+    testData.map(predict)
+
+  /**
+   * Predict labels for provided features
+   * Using a piecewise constant function
+   *
+   * @param testData features to be labeled
+   * @return predicted labels
+   */
+  def predict(testData: JavaRDD[java.lang.Double]): JavaDoubleRDD =
+    JavaDoubleRDD.fromRDD(predict(testData.rdd.asInstanceOf[RDD[Double]]))
+
+  /**
+   * Predict a single label
+   * Using a piecewise constant function
+   *
+   * @param testData feature to be labeled
+   * @return predicted label
+   */
+  def predict(testData: Double): Double = {
+    val result = binarySearch(features, testData)
+
+    val index =
+      if (result == -1) {
```

Comment:
As for the special singularity case I believe this requires further considerations. Currently we just sort the input to PAV by feature therefore order of multiple data points with the same feature is undefined.

Consider a case where features are 1, 2, 2, 3 and labels are in first case 1, 4, 2, 5 and in second case 1, 2, 4, 5. For first case the result of PAV would be 1, 3, 3, 5 but in second case 1, 2, 4, 5.

Similarly for `IsotonicRegressionModel` with boundaries 1, 2, 2, 3 and predictions in first case 1, 4, 2, 5 and in second case 1, 2, 4, 5. Now the first mode would return predict(1.5)=2.5, predict(2.5)=3.5, but the second would return 1.5 and 4.5 respectively for the same input values.

I suggest to sort the input by features and then by labels if features are equal. The same would be true for the model. Therefore both PAV and the predictions of values between boundaries would be deterministic. The predictions for the boundary with multiple values would remain non-deterministic (based on `Java.util.Arrays.binarySearch()` which in this case also returns one of the correct results, but does not specify which).